include/nbl/builtin/hlsl/bxdf/ndf/ggx.hlsl

```diff
@@ -132,56 +143,88 @@ struct GGX<T,true NBL_PARTIAL_REQ_BOT(concepts::FloatingPointScalar<T>) >
     }
 
-    scalar_type G1_wo_numerator(scalar_type NdotX, scalar_type devsh_part)
+    scalar_type G1_wo_numerator_devsh_part(scalar_type NdotX, scalar_type devsh_part)
     {
         return scalar_type(1.0) / (NdotX + devsh_part);
     }
 
-    scalar_type correlated_wo_numerator(scalar_type ax2, scalar_type ay2, scalar_type NdotV, scalar_type TdotV2, scalar_type BdotV2, scalar_type NdotV2, scalar_type NdotL, scalar_type TdotL2, scalar_type BdotL2, scalar_type NdotL2)
+    scalar_type correlated_wo_numerator(scalar_type absNdotV, scalar_type TdotV2, scalar_type BdotV2, scalar_type NdotV2, scalar_type absNdotL, scalar_type TdotL2, scalar_type BdotL2, scalar_type NdotL2)
     {
-        scalar_type Vterm = NdotL * devsh_part(TdotV2, BdotV2, NdotV2, ax2, ay2);
-        scalar_type Lterm = NdotV * devsh_part(TdotL2, BdotL2, NdotL2, ax2, ay2);
+        scalar_type Vterm = absNdotL * devsh_part(TdotV2, BdotV2, NdotV2);
+        scalar_type Lterm = absNdotV * devsh_part(TdotL2, BdotL2, NdotL2);
         return scalar_type(0.5) / (Vterm + Lterm);
     }
 
-    scalar_type G2_over_G1(scalar_type ax2, scalar_type ay2, bool transmitted, scalar_type NdotV, scalar_type TdotV2, scalar_type BdotV2, scalar_type NdotV2, scalar_type NdotL, scalar_type TdotL2, scalar_type BdotL2, scalar_type NdotL2)
+    scalar_type G2_over_G1(bool transmitted, scalar_type NdotV, scalar_type TdotV2, scalar_type BdotV2, scalar_type NdotV2, scalar_type NdotL, scalar_type TdotL2, scalar_type BdotL2, scalar_type NdotL2)
     {
         scalar_type G2_over_G1;
         if (transmitted)
         {
-            scalar_type devsh_v = devsh_part(TdotV2, BdotV2, NdotV2, ax2, ay2);
-            scalar_type L_v = scalar_type(0.5) * (devsh_v / NdotV - scalar_type(1.0));
-            scalar_type L_l = scalar_type(0.5) * (devsh_part(TdotL2, BdotL2, NdotL2, ax2, ay2) / NdotL - scalar_type(1.0));
-            G2_over_G1 = hlsl::beta<scalar_type>(scalar_type(1.0) + L_l, scalar_type(1.0) + L_v);
-            G2_over_G1 *= scalar_type(1.0) + L_v;
+            if (NdotV < 1e-7 || NdotL < 1e-7)
+                return 0.0;
+            scalar_type onePlusLambda_V = scalar_type(0.5) * (devsh_part(TdotV2, BdotV2, NdotV2) / NdotV + scalar_type(1.0));
+            scalar_type onePlusLambda_L = scalar_type(0.5) * (devsh_part(TdotL2, BdotL2, NdotL2) / NdotL + scalar_type(1.0));
+            G2_over_G1 = hlsl::beta<scalar_type>(onePlusLambda_L, onePlusLambda_V) * onePlusLambda_V;
         }
         else
         {
-            scalar_type devsh_v = devsh_part(TdotV2, BdotV2, NdotV2, ax2, ay2);
+            scalar_type devsh_v = devsh_part(TdotV2, BdotV2, NdotV2);
             G2_over_G1 = NdotL * (devsh_v + NdotV);
-            G2_over_G1 /= NdotV * devsh_part(TdotL2, BdotL2, NdotL2, ax2, ay2) + NdotL * devsh_v;
+            G2_over_G1 /= NdotV * devsh_part(TdotL2, BdotL2, NdotL2) + NdotL * devsh_v;
         }
 
         return G2_over_G1;
     }
```

Comment:
same comments as isotropic case, every `Ndot` everywhere needs to be properly prefixed
- abs (cannot be negative and expected to not be clamped)
- clamped (cannot be negative but is clamped
- no prefix (can be negative)

IMHO this is why its important to have and use `getClampedXdotY(clamp_mode)` on the interactions, caches etc.

Reply:
We could also solve this by having and using `getClampedXdotY(clamp_mode)` on the interactions, caches etc.

then the functions have smaller signatures as well (but are templated)